lading/src/blackhole/datadog_stateful_logs.rs

```
@@ -0,0 +1,225 @@
+pub struct Config {
+    /// The binding address for the gRPC server
+    pub grpc_addr: SocketAddr,
+    /// Delay to add before sending responses (in milliseconds)
+    #[serde(default = "default_response_delay_millis")]
+    pub response_delay_millis: u64,
+}
+
+fn default_response_delay_millis() -> u64 {
+    0
+}
+
+#[derive(Debug)]
+/// The Datadog stateful logs gRPC blackhole.
+pub struct DatadogStatefulLogs {
+    grpc_addr: SocketAddr,
+    shutdown: lading_signal::Watcher,
+    metric_labels: Vec<(String, String)>,
+    response_delay: Duration,
+}
+
+impl DatadogStatefulLogs {
+    /// Create a new [`DatadogStatefulLogs`] server instance
+    #[must_use]
+    pub fn new(general: General, config: Config, shutdown: lading_signal::Watcher) -> Self {
+        let mut metric_labels = vec![
+            ("component".to_string(), "blackhole".to_string()),
+            (
+                "component_name".to_string(),
+                "datadog_stateful_logs".to_string(),
+            ),
+        ];
+        if let Some(id) = general.id {
+            metric_labels.push(("id".to_string(), id));
+        }
+
+        Self {
+            grpc_addr: config.grpc_addr,
+            shutdown,
+            metric_labels,
+            response_delay: Duration::from_millis(config.response_delay_millis),
+        }
+    }
+
+    /// Run [`DatadogStatefulLogs`] to completion
+    ///
+    /// This function runs the server forever, unless a shutdown signal is
+    /// received or an unrecoverable error is encountered.
+    ///
+    /// # Errors
+    ///
+    /// Function will return an error if the server fails to start.
+    pub async fn run(self) -> Result<(), Error> {
+        let mut join_set = JoinSet::new();
+
+        let grpc_task = run_server(self.grpc_addr, self.response_delay, &self.metric_labels);
+        join_set.spawn(async move {
+            grpc_task.await.unwrap_or_else(|e| {
+                error!("gRPC task failed: {}", e);
+            });
```

This `unwrap_or_else` swallows `e`, meaning we only ever log errors and do nothing about them. Just to confirm that logic is intentional. Based on L115 I think it must be, and is acceptable.

---

Modified the logic to error out of `run` if the grpc server errors or ends before lading termination.